src/Text/Parsing/Parser/Combinators.purs

```diff
@@ -200,27 +200,33 @@ try p = (ParserT <<< ExceptT <<< StateT) \(s@(ParseState _ _ consumed)) -> do
 -- | >>> runParser "ac" (tryRethrow (char 'a' *> char 'b'))
 -- | Left (ParseError "Expected 'b'" (Position { line: 1, column: 1 }))
 -- | ```
-tryRethrow :: forall m s a. Monad m => ParserT s m a -> ParserT s m a
-tryRethrow p = (ParserT <<< ExceptT <<< StateT) \(s@(ParseState _ position consumed)) -> do
-  Tuple e s'@(ParseState input' position' _) <- runStateT (runExceptT (unwrap p)) s
-  case e of
-    Left (ParseError err _) -> pure (Tuple (Left (ParseError err position)) (ParseState input' position' consumed))
-    _ -> pure (Tuple e s')
+tryRethrow :: forall m s a. ParserT s m a -> ParserT s m a
+tryRethrow (ParserT k1) = ParserT
+  ( mkFn5 \state1@(ParseState _ position consumed) more lift throw done ->
+      runFn5 k1 state1 more lift
+        ( mkFn2 \(ParseState input' position' _) (ParseError err _) ->
+            runFn2 throw (ParseState input' position' consumed) (ParseError err position)
+        )
+        done
+  )
 
 -- | Parse a phrase, without modifying the consumed state or stream position.
-lookAhead :: forall s a m. Monad m => ParserT s m a -> ParserT s m a
-lookAhead p = (ParserT <<< ExceptT <<< StateT) \s -> do
-  Tuple e _ <- runStateT (runExceptT (unwrap p)) s
-  pure (Tuple e s)
+lookAhead :: forall s a m. ParserT s m a -> ParserT s m a
+lookAhead (ParserT k1) = ParserT
+  ( mkFn5 \state1 more lift throw done ->
+      runFn5 k1 state1 more lift
+        (mkFn2 \_ err -> runFn2 throw state1 err)
+        (mkFn2 \_ res -> runFn2 done state1 res)
+  )
 
 -- | Match one or more times.
-many1 :: forall m s a. Monad m => ParserT s m a -> ParserT s m (NonEmptyList a)
+many1 :: forall m s a. ParserT s m a -> ParserT s m (NonEmptyList a)
 many1 p = NEL.cons' <$> p <*> many p
 
 -- | Match one or more times.
 -- |
 -- | Stack-safe version of `many1` at the expense of a `MonadRec` constraint.
```

Comment:
Not necessarily, or at least I don't think you should just remove the Rec implementations without testing. The parser is always stack-safe now, yes, but the iterative implementations may be more efficient at runtime. For example, the upstream `many` combinator is significantly slower than `manyRec`. It may be the case that we should copy over the Rec implementations and remove the calls to tailRecM, just using standard Monadic recursion. It could also be the case that the "naive" implementations perform fine in comparison and we should drop the Rec implementaiton.